python/oneseismic/scan/scanners.py

```diff
@@ -206,3 +207,201 @@ def report(self):
+
+
+    def report(self):
+        # To represent a point (xline, iline) as UTM (x, y) coordinate only a
+        # combination of linear transformations is needed inside one UTM zone
+        # grid, because converting from one cartesian grid to another is always
+        # a linear transformation.
+        #
+        # Thus a problem of changing coordinate system from iline/xline to UTM
+        # coordinates presents a standard 2D affine transformation:
+        #   a b c     p.x     p.inline
+        #   k m n  x  p.y  =  p.xline
+        #   0 0 1     1       1
+        #
+        # As we have 3 variables for each equation we need to obtain 3 points
+        # with known coordinates in each of the coordinate systems. Then we can
+        # solve the system with regards to coefficient matrix.
+        #
+        # With 3 known points system can be represented as
+        #   a b c     p0.x  p1.x  p2.x     p0.iline  p1.iline  p2.iline
+        #   k m n  x  p0.y  p1.y  p2.y  =  p0.xline  p1.xline  p2.xline
+        #   0 0 1     1     1     1        1         1         1
+        #
+        # This can be transformed to the system M x = b, where
+        #
+        #       p0.x  p0.y  1       a k       p0.inline  p0.xline
+        #   M = p1.x  p1.y  1   x = b m   b = p1.inline  p1.xline
+        #       p2.x  p2.y  1 ,     c n,      p2.inline  p2.xline
+        #
+        # We solve this system using the inverse of M x = M^-1 b
+
+        M = np.array([
+            [self.p0[0], self.p0[1], 1],
+            [self.p1[0], self.p1[1], 1],
+            [self.p2[0], self.p2[1], 1]
+        ])
+
+        try:
+            utm_to_inline = np.linalg.inv(M).dot(
+                np.array([
+                    self.p0_linenos[0],
+                    self.p1_linenos[0],
+                    self.p2_linenos[0]
+                ])
+            )
+            utm_to_crossline = np.linalg.inv(M).dot(
+                np.array([
+                    self.p0_linenos[1],
+                    self.p1_linenos[1],
+                    self.p2_linenos[1]
+                ])
+            )
+        except np.linalg.LinAlgError:
+            # This exception is raised if M is not invertible
+            # TODO: Raise exception? Log?
+            return {}
+
+        # Check that the lineno-to-utm mappings are not zero. This will for
+        # instance happens if the UTM coordinate headers are set to zero
+        # (happens if they are unset).
+        # TODO: Raise exception? Log?
+        if (utm_to_inline == 0).all() or (utm_to_crossline == 0).all():
+            return {}
+
+        # We expect inlines and crosslines to be perpendicular to each other.
+        # If that is the case, the vectors (a, b) and (k, m) will also be
+        # perpendicular, a, b, k and m being coefficients on the UTM-to-lineno
+        # affine matrix (see comment at beginning of function). We divide by the
+        # product of the moduli of the two vectors to distinguish between
```

Comment:
Can you tell me why??
Like I see the perpendicular step. I see the normalization step - separately.
But why is the way you combine them correct? I can't find the math behind your moduli product and dot/moduli_product division, so I don't really know if you check makes sense. What is it that you are doing?

To me it would make more sense to first normalize vectors and then check if they are perpendicular.